### PR TITLE
Infer bounds for Reduce op

### PR DESCRIPTION
**Pull request by smit-hinsu**
Extended reduce op shape function to also propagate bounds for dimensions that are not reduced.

Also, moved verifyCompatibleShapeWithBounds to public in Base.h

cc @zhouxin913  for review

### CODE REVIEW

**Comment by zhouxin913** [comment]
Hi Smit, as I have just submit https://github.com/openxla/stablehlo/pull/401, this PR will have conflict with HEAD. You may consider rebase now or any time later: my PR only moves some checks in a util function (For `ReduceOp` it is `verifyReduceOpInputsAndInferShape()` without any logical change.
This is not a blocker, I will continue my review and leave comments, just FYI.

**Comment by smit-hinsu** [comment]
> Hi Smit, as I have just submit #401, this PR will have conflict with HEAD. You may consider rebase now or any time later: my PR only moves some checks in a util function (For `ReduceOp` it is `verifyReduceOpInputsAndInferShape()` without any logical change. This is not a blocker, I will continue my review and leave comments, just FYI.

Rebased this.

**Review by zhouxin913 — COMMENTED** [review]
LGTM with a tiny comment in tests.

Thanks for your fix in https://reviews.llvm.org/D139271: with which we can add encoding when init inferred type even with `ShapedTypeComponents` interface.

Besides, this PR makes issue https://github.com/openxla/stablehlo/issues/394 more complicated, which can be addressed at that issue (I have left a comment there, so that this PR is good to go)
Assign to @burmako for a second review.

File: stablehlo/tests/infer_stablehlo.mlir [hunk 1]
```diff
@@ -600,21 +600,63 @@ func.func @complex_sparsity(%arg0: tensor<10x10xf32, #CSR>, %arg1: tensor<10x10x
...
+  %2 = "hlo_test_infer.get_return_types"(%0)
+      : (tensor<?x5xf32, #stablehlo.type_extensions<bounds = [7, ?]>>) -> tensor<*xindex>
+
+  func.return %2: tensor<*xindex>
+}
+
+// -----
+
+// CHECK-LABEL: func @unranked_reduce
+func.func @unranked_reduce(%arg0: tensor<*xf32>, %arg1 : tensor<f32>)
```
[the remark this answers is not in the document]
Comment (smit-hinsu) [reply stablehlo/tests/infer_stablehlo.mlir:645]: yes, this is not related to bounds but we still need this as it improves code coverage in general. Could move to a separate PR.

That test doesn't test allInputsUnranked case but reduce_single_operand_unranked does.

However that also can't test if the inferred result for unranked inputs is actually unranked. It should pass as long as the result is compatible with unranked. Basically, any result with the right element type should pass that.

[the remark this answers is not in the document]
Comment (zhouxin913) [reply stablehlo/tests/infer_stablehlo.mlir:645]: "that also can't test if the inferred result for unranked inputs is actually unranked." True, thanks for this! Feel feel to keep it here as well. Resolved.

**Review by burmako — APPROVED** [review]

**Pull request merged**